
<launch>
   
   <!-- RGB-D MAPPING VERSION -->
   <!-- WARNING : Database is automatically deleted on each startup -->
   <!--           See "delete_db_on_start" option below... -->
  
   <!-- ODOMETRY ARGUMENTS: "strategy", "feature", "nn" and "local_map":
        -Strategy: 0=BOW (bag-of-words) 1=Optical Flow
        -Feature type: 0=SURF 1=SIFT 2=ORB 3=FAST/FREAK 4=FAST/BRIEF 5=GFTT/FREAK 6=GFTT/BRIEF 7=BRISK
        -Nearest neighbor strategy : 0=Linear, 1=FLANN_KDTREE, 2=FLANN_LSH, 3=BRUTEFORCE 
           Set to 1 for float descriptor like SIFT/SURF                  
           Set to 3 for binary descriptor like ORB/FREAK/BRIEF/BRISK      
        -Local map size: number of unique features to keep track 
    -->
   <arg name="strategy" default="0" />
   <arg name="feature" default="6" />
   <arg name="nn" default="3" />
   <arg name="local_map" default="1000" />
   
  <!-- TF FRAMES -->
  <node pkg="tf" type="static_transform_publisher" name="base_to_camera_tf" 
    args="0.0 0.0 0.0 0.0 0.0 0.0 /base_link /camera_link 100" />
     
  <group ns="rtabmap">
  
    <!-- Odometry -->
    <node pkg="rtabmap" type="rgbd_odometry" name="visual_odometry" output="screen">
      <remap from="rgb/image"       to="/camera/rgb/image_rect_color"/>
      <remap from="depth/image"     to="/camera/depth_registered/image_raw"/>
      <remap from="rgb/camera_info" to="/camera/depth_registered/camera_info"/>
	  
	  <param name="Odom/Strategy" type="string" value="$(arg strategy)"/> 
	  <param name="Odom/FeatureType" type="string" value="$(arg feature)"/>            
      <param name="OdomBow/NNType" type="string" value="$(arg nn)"/>
      <param name="OdomBow/LocalHistorySize" type="string" value="$(arg local_map)"/>
   
      <param name="frame_id" type="string" value="base_link"/>
    </node>
  
    <!-- Visual SLAM (robot side) -->
    <!-- args: "delete_db_on_start" and "udebug" -->
    <node name="rtabmap" pkg="rtabmap" type="rtabmap" output="screen" args="--delete_db_on_start">
      <param name="subscribe_depth" type="bool" value="true"/>
      <param name="subscribe_laserScan" type="bool" value="false"/>
	
      <remap from="rgb/image" to="/camera/rgb/image_rect_color"/>
      <remap from="depth/image" to="/camera/depth_registered/image_raw"/>
      <remap from="rgb/camera_info" to="/camera/depth_registered/camera_info"/>
      <remap from="odom" to="odom"/>
      
      <param name="RGBD/LocalLoopDetectionTime" type="string" value="false"/>
      <param name="LccBow/MinInliers" type="string" value="10"/>
      <param name="LccBow/InlierDistance" type="string" value="0.02"/>
      	
      <param name="queue_size" type="int" value="30"/>
    </node>
  
    <!-- Visualisation (client side) -->
    <node pkg="rtabmap" type="rtabmapviz" name="rtabmapviz" args="-d $(find rtabmap)/launch/config/rgbd_gui.ini" output="screen">
      <param name="subscribe_depth" type="bool" value="true"/>
      <param name="subscribe_laserScan" type="bool" value="false"/>
      <param name="queue_size" type="int" value="30"/>
    
      <remap from="rgb/image" to="/camera/rgb/image_rect_color"/>
      <remap from="depth/image" to="/camera/depth_registered/image_raw"/>
      <remap from="rgb/camera_info" to="/camera/depth_registered/camera_info"/>
      <remap from="odom" to="odom"/>
    </node>
  
  </group>

</launch>
